### ISSUE TITLE
Add Support for Solr Authentication

### ISSUE THREAD

**Issue by tdonohue** [DSpace/DSpace#3169]
Currently, we assume that Solr is accessible to the REST API / backend without any authentication. We should consider minimally supporting Solr's Basic Authentication for anyone who wishes to require authentication into Solr.

https://lucene.apache.org/solr/guide/8_7/basic-authentication-plugin.html

Solr also supports a number of other Authentication Plugins, but it appears most plugins are not yet supported in SolrJ (which we use). 

**Workaround** 

Currently, if you install Solr with Basic Authentication enabled, you'll end up seeing `Error 401 require authentication` errors from Solr.  If you encounter that, the workaround is to set `AuthType="None"` in your `solr.conf`.  **Just be sure that your Solr is secured in another way, often behind a firewall.  The Solr maintainers recommend that Solr never be publicly available on the web.**

**Comment by hardyoyo** [comment]
some Solr installations utilize a solr.in.sh script to set options as environment variables. A possible location for such a script on an RHEL/CentOS/AmazonLinux machine will be at /etc/default/solr.in.sh

**Comment by hardyoyo** [comment]
In reading more about the Basic Auth plugin for Solr, it seems you can

> use SolrJ’s PreemptiveBasicAuthClientBuilderFactory to add basic authentication credentials to all requests automatically

https://solr.apache.org/guide/8_1/basic-authentication-plugin.html#global-jvm-basic-auth-credentials

I.e. it may be possible to hard wire the credentials for all requests... Now, if you're wondering what the point of that exercise would be, I join you in wondering. But, it's definitely worth checking out.

**Comment by tdonohue** [comment]
Moving this to the 7.2 board, as this is very low priority (until we have users requesting this as a feature).

**Comment by tehaef** [comment]
> > use SolrJ’s PreemptiveBasicAuthClientBuilderFactory to add basic authentication credentials to all requests automatically

After experimenting with this authentication method, i got the impression that it does not work in DSpace 9.2 out of the box. If i understand the code correctly, DSpace initializes the HttpSolrClient with its own httpConnectionPoolService, that is built ontop of Apache HttpClient library, thus ignoring the previously provided credentials; and this client doesn't seem to support a similar feature (s.a.: https://hc.apache.org/httpcomponents-client-4.5.x/current/tutorial/html/authentication.html#d5e717). Am i right under the assumption, that thus, this kind of Solr authentication will currently not work in DSpace?

**Comment by tdonohue** [comment]
@tehaef : Correct, this ticket is still open and unclaimed by a volunteer.  So, it appears that Solr Authentication still is not supported by DSpace.

This ticket still is looking for a volunteer to investigate how to implement this.

**Comment by JohnnyMendesC** [comment]
Hi, @lgeggleston @tdonohue @tehaef @hardyoyo 
I took a look at this and opened issue #12982

To answer @tehaef’s question above: yes, the `PreemptiveBasicAuthClientBuilderFactory` doesn’t seem to help in this case. From what I’ve found, DSpace never lets SolrJ build its own `HttpClient`, every Solr client (search, statistics, authority, oai, qaevent, suggestion, audit, as well as some CLI tools) are constructed with an explicit `HttpClient`/ `HttpSolrClient` that DSpace builds itself and passes via `.withHttpClient(...)`, so that system property is never consulted.

The PR instead adds a small preemptive Basic-Auth interceptor, configured via a shared and optional `solr.authentication.*` config block that applies by default across all Solr cores (since the Solr basic authentication plugin protects the whole instance, not individual cores) with a couple of cores that can optionally point at a separate Solr instance getting their own override.

Feedback on the overall approach is very welcome. (_I'm still working on a new test `SolrAuthUtilsTest.java` file at the moment too_)

_Johnny Mendes - Acervos Action-IT_

**Comment by JohnnyMendesC** [comment]
Update on #12982. 
Added test coverage since my last comment:

- `SolrAuthUtilsTest` (unit tests): covers the config parsing/validation logic, missing `type`, `type=none`, an unrecognized `type` value, a missing `user`, the fallback-prefix behavior (e.g. `iiif.search.authentication.*` falling back to `solr.authentication.*`), and that the interceptor is only attached when authentication is actually configured.
- An integration test added to the existing `HttpConnectionPoolServiceTest`, asserting the `Authorization` header is actually present on an outgoing request once `solr.authentication.*` is set (using a MockServer expectation on the header itself, building on the same harness already used by that test class).

I haven't yet had the chance to perform a final manual check on a real Solr instance with `security.json`/`BasicAuthPlugin` enabled in my local setup. But with the new unit and integration tests ready and passing, I’ll mark this as ready for review. If anyone has the opportunity to perform a manual check too, it will be very valuable. In the meantime, I remain very open to feedback on the overall approach.

_Johnny Mendes - Acervos Action-IT_
